media/com_jedchecker/js/script.js

```diff
@@ -1,92 +1,105 @@
-
-			target.html('<span class="text-danger"><b>' + status + ':</b> ' + xhr.status + ' ' + xhr.statusText + '</span>');
-
-			sidebar.find('.badge.bg-danger').text('?');
-			sidebar.find('.badge.bg-warning,.badge.bg-secondary,.badge.bg-info').text('');
-			sidebar.find('.badge.bg-success').addClass('hidden');
-
-			sidebar.find('.spinner-border').addClass('hidden');
-		}
-	});
-}
-
-var jed_collapse_init = false;
-Joomla.submitbutton = function (task) {
-	if (task == 'check') {
-		jQuery(".jedchecker-results").removeClass("hidden");
-		jQuery('.jedchecker-results .badge:not(.bg-success)').html('');
-		jQuery('.jedchecker-results .badge.bg-success').addClass('hidden');
-		jQuery('.jedchecker-results .spinner-border').removeClass('hidden');
-		jQuery('.police-check-result').html('<div class="text-center text-info"><span class="spinner-border"></span></div>');
-
-		if (!jed_collapse_init) {
-			jQuery(".card-header[data-bs-toggle]")
-				.addClass("accordion-button collapsed")
-				.each(function () {
-					var $this = jQuery(this);
-					$this.attr('href', $this.attr('data-href'));
-				});
-
-			new bootstrap.Collapse(document.getElementById('jedchecker-welcome'));
-			new bootstrap.Collapse(document.getElementById('jedchecker-contributors'));
-
-			jed_collapse_init = true;
-		}
-
-		for (index = 0; index < jed_options["rules"].length; ++index) {
-			check(jed_options["url"], jed_options["rules"][index]);
-		}
-
-	} else {
-		Joomla.submitform(task);
-	}
-}
-
-jQuery(document).ready(function() {
-	new bootstrap.Tooltip(document.getElementById('jedchecker'), {container: 'body', selector: '[data-bs-toggle=tooltip]'});
-});
+(() => {
+  const jsonDataElement = document.getElementById('jed-rules-json');
+  if (!jsonDataElement) {
+    throw new Error('Initialization data is missing');
+  }
+  let jedOptions;
+  try {
+    jedOptions = JSON.parse(jsonDataElement.innerHTML);
+  } catch (e) {
+    throw new Error('Initialization data is missing');
+  }
+
+  if (!jedOptions) throw new Error('Initialization data is missing');
+
+  add_validation = () => {
+    // Loop over them and prevent submission
+    [...document.querySelectorAll('.needs-validation')].forEach((form) => {
+      form.addEventListener('submit', (event) => {
+        const form = event.target;
+        if (form.checkValidity() === false) {
+          event.preventDefault();
+          event.stopPropagation();
+        }
+        form.classList.add('was-validated');
+      }, false);
+    });
+  }
+
+  function check(url, rule) {
+    fetch(`${url}index.php?option=com_jedchecker&task=police.check&format=raw&rule=${rule}`)
+      .then(response => response.text())
+      .then(data => {
+        const sidebar = document.querySelector(`#jed-${rule}`);
+        const target = document.querySelector(`#police-check-result-${rule}`);
+
+        target.innerHTML = data;
+
+        const error = [...target.querySelectorAll('.alert-danger')].length;
```

Comment:
It's possible to get length directly (without the spread operator), i.e. `target.querySelectorAll('.alert-danger').length`. Same in other places.

Reply:
No, the `querySelectorAll` returns either a node list or undefined, so if you do `.length` directly there and something goes wrong in the HTML part then you might end up doing `undefined.length`. Always make the node lists an array so even if it's empty nothing throws but most importantly you can use the array methods directly

Reply:
Are you sure? I've never met `querySelectorAll` returning undefined for correct query string (and even for an incorrect one it  throws `DOMException` instead).

Reply:
Yes I'm sure: https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll#return_value but you can test it in your console

Reply:
They say it's always `NodeList`: "A non-live NodeList containing one Element object for each element that matches at least one of the specified selectors or an empty NodeList in case of no matches."

Reply:
That’s the specs real life is bit different there are still browsers out there that return undefined. Also there’s no perf gain here by not transforming the node list to an array only degraded dx. As the Joomla js leader I always recommended transforming the node lists to arrays so no weird surprises for the next devs that will try to enhance the code…

You are welcome to change the code by I won’t